### ISSUE TITLE
Argument refurbishing breaks cgo

### ISSUE THREAD

**Issue by shulima** [Homebrew/homebrew#45657]
I'm (re)attempting to produce a formula for libguestfs. I've been struggling to build Go bindings because of the 'go install' called by libguestfs Makefile failing with a cryptic message:

```
Making all in golang
 cd .. && /bin/sh /private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/build-aux/missing automake-1.15 --foreign golang/Makefile
 cd .. && /bin/sh ./config.status golang/Makefile
config.status: creating golang/Makefile
../run go install libguestfs.org/guestfs
# libguestfs.org/guestfs
cannot load DWARF output from $WORK/libguestfs.org/guestfs/_obj//_cgo_.o: decoding dwarf section info at offset 0x0: too short
make[2]: *** [pkg/darwin_amd64/libguestfs.org/guestfs.a] Error 2
```

After dropping to post-BuildError shell, the same command would work just fine. 

After much debugging, I've noticed that `$HOMEBREW_CCCFG` is set to `Osa` during the build, but to `sa` when dropped to the interactive shell. I've managed to reproduce the issue by temporarily overriding the variable. See below for behaviour without and with the `O` flag (I've added the -x option to `go install` to make it print out commands used).
```
bash-3.2$ echo $HOMEBREW_CCCFG
sa
bash-3.2$ ../run go install -x libguestfs.org/guestfs
WORK=/var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build731081957
mkdir -p $WORK/libguestfs.org/guestfs/_obj/
mkdir -p $WORK/libguestfs.org/
cd /private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/golang/src/libguestfs.org/guestfs
CGO_LDFLAGS="-L/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src/.libs" "-lguestfs" /usr/local/Cellar/go/1.5.1/libexec/pkg/tool/darwin_amd64/cgo -objdir $WORK/libguestfs.org/guestfs/_obj/ -importpath libguestfs.org/guestfs -- -I $WORK/libguestfs.org/guestfs/_obj/ -I/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src -DGUESTFS_PRIVATE=1 guestfs.go
clang -I . -fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common -print-libgcc-file-name
clang -I . -fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common -I $WORK/libguestfs.org/guestfs/_obj/ -I/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src -debug-gcc -DGUESTFS_PRIVATE=1 -o $WORK/libguestfs.org/guestfs/_obj/_cgo_main.o -c $WORK/libguestfs.org/guestfs/_obj/_cgo_main.c
clang -I . -fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common -I $WORK/libguestfs.org/guestfs/_obj/ -I/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src -debug-gcc -DGUESTFS_PRIVATE=1 -o $WORK/libguestfs.org/guestfs/_obj/_cgo_export.o -c $WORK/libguestfs.org/guestfs/_obj/_cgo_export.c
clang -I . -fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common -I $WORK/libguestfs.org/guestfs/_obj/ -I/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src -debug-gcc -DGUESTFS_PRIVATE=1 -o $WORK/libguestfs.org/guestfs/_obj/guestfs.cgo2.o -c $WORK/libguestfs.org/guestfs/_obj/guestfs.cgo2.c
clang -I . -fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common -o $WORK/libguestfs.org/guestfs/_obj/_cgo_.o $WORK/libguestfs.org/guestfs/_obj/_cgo_main.o $WORK/libguestfs.org/guestfs/_obj/_cgo_export.o $WORK/libguestfs.org/guestfs/_obj/guestfs.cgo2.o -L/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src/.libs -lguestfs
/usr/local/Cellar/go/1.5.1/libexec/pkg/tool/darwin_amd64/cgo -objdir $WORK/libguestfs.org/guestfs/_obj/ -dynpackage guestfs -dynimport $WORK/libguestfs.org/guestfs/_obj/_cgo_.o -dynout $WORK/libguestfs.org/guestfs/_obj/_cgo_import.go
clang -I . -fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common -o $WORK/libguestfs.org/guestfs/_obj/_all.o $WORK/libguestfs.org/guestfs/_obj/_cgo_export.o $WORK/libguestfs.org/guestfs/_obj/guestfs.cgo2.o -L/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src/.libs -Wl,-r -nostdlib
/usr/local/Cellar/go/1.5.1/libexec/pkg/tool/darwin_amd64/compile -o $WORK/libguestfs.org/guestfs.a -trimpath $WORK -p libguestfs.org/guestfs -buildid 0d93b59a16cb23daaee27febcfb7c0fefbd57bb0 -D _/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/golang/src/libguestfs.org/guestfs -I $WORK -pack $WORK/libguestfs.org/guestfs/_obj/_cgo_gotypes.go $WORK/libguestfs.org/guestfs/_obj/guestfs.cgo1.go $WORK/libguestfs.org/guestfs/_obj/_cgo_import.go
pack r $WORK/libguestfs.org/guestfs.a $WORK/libguestfs.org/guestfs/_obj/_all.o # internal
mkdir -p /private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/golang/pkg/darwin_amd64/libguestfs.org/
mv $WORK/libguestfs.org/guestfs.a /private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/golang/pkg/darwin_amd64/libguestfs.org/guestfs.a
```
```
bash-3.2$ HOMEBREW_CCCFG=Osa ../run go install -x libguestfs.org/guestfs
WORK=/var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build866479474
mkdir -p $WORK/libguestfs.org/guestfs/_obj/
mkdir -p $WORK/libguestfs.org/
cd /private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/golang/src/libguestfs.org/guestfs
CGO_LDFLAGS="-L/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src/.libs" "-lguestfs" /usr/local/Cellar/go/1.5.1/libexec/pkg/tool/darwin_amd64/cgo -objdir $WORK/libguestfs.org/guestfs/_obj/ -importpath libguestfs.org/guestfs -- -I $WORK/libguestfs.org/guestfs/_obj/ -I/private/tmp/libguestfs20151103-23166-ogi40h/libguestfs-1.30.3/src -DGUESTFS_PRIVATE=1 guestfs.go
# libguestfs.org/guestfs
cannot load DWARF output from $WORK/libguestfs.org/guestfs/_obj//_cgo_.o: decoding dwarf section info at offset 0x0: too short
```

Here's post-build-error configuration/ENV:
```
==> Formula
Tap: homebrew/homebrew-fuse
Path: /usr/local/Library/Taps/homebrew/homebrew-fuse/Formula/libguestfs.rb
==> Configuration
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: 959922c4aa2d0165a1054796bceb7ff99782aba5
Last commit: 10 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: quad-core 64-bit haswell
OS X: 10.10.5-x86_64
Xcode: 7.0.1
CLT: 7.1.0.0.1.1444952191
Clang: 7.0 build 700
X11: 2.7.8 => /opt/X11
System Ruby: 2.0.0-p481
Perl: /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/local/bin/ruby => /usr/local/Cellar/ruby/2.2.0/bin/ruby
Java: 1.8.0_31
==> ENV
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
CMAKE_PREFIX_PATH: /usr/local/opt/gettext:/usr/local/opt/libffi:/usr/local/opt/readline:/usr/local/opt/libxml2:/usr/local
CMAKE_INCLUDE_PATH: /System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
CMAKE_LIBRARY_PATH: /System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
PKG_CONFIG_PATH: /usr/local/opt/libtasn1/lib/pkgconfig:/usr/local/opt/nettle/lib/pkgconfig:/usr/local/opt/gnutls/lib/pkgconfig:/usr/local/opt/libffi/lib/pkgconfig:/usr/local/opt/gli
b/lib/pkgconfig:/usr/local/opt/pixman/lib/pkgconfig:/usr/local/opt/xz/lib/pkgconfig:/usr/local/opt/libxml2/lib/pkgconfig:/usr/local/opt/augeas/lib/pkgconfig:/usr/local/opt/pcre/lib/
pkgconfig:/usr/local/opt/yajl/share/pkgconfig:/usr/local/Library/ENV/pkgconfig/fuse
PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Library/ENV/pkgconfig/10.10
ACLOCAL_PATH: /usr/local/opt/gettext/share/aclocal:/usr/local/opt/libxml2/share/aclocal:/usr/local/share/aclocal
PATH: /usr/local/Library/ENV/4.3:/usr/local/opt/go/bin:/usr/local/opt/autoconf/bin:/usr/local/opt/automake/bin:/usr/local/opt/libtool/bin:/usr/local/opt/pkg-config/bin:/usr/local/op
t/jpeg/bin:/usr/local/opt/libtasn1/bin:/usr/local/opt/nettle/bin:/usr/local/opt/gnutls/bin:/usr/local/opt/gettext/bin:/usr/local/opt/glib/bin:/usr/local/opt/qemu/bin:/usr/local/opt/
xz/bin:/usr/local/opt/yajl/bin:/usr/local/opt/cdrtools/bin:/usr/local/opt/libxml2/bin:/usr/local/opt/augeas/bin:/usr/local/opt/pcre/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

I'm happy to provide more info/output as requested (my debugging capabilities have been exhausted).

**Comment by mikemcquaid** [comment]
Try the requested steps on the Troubleshooting page: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Troubleshooting.md#troubleshooting

Please follow (all of) these steps and post the information here so we can help you with your problem.

Thanks!

**Comment by DomT4** [comment]
The CLT/Xcode mismatch may be problematic here.

**Comment by shulima** [comment]
Apologies for not looking at the Troubleshooting page first (when everything else fails, read the documentation). If updating Xcode had helped, I'd be very ashamed right now.

Updated Xcode and rerun brew update/doctor, the issue persists.

Uploaded gist-logs: https://gist.github.com/732352d7c7a27d8375c3
(They're very enlightening. Thanks again for pointing me at the Troubleshooting page.)

This looks fishy:
```
clang called with: -w -Wno-error -o/var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build437084003/libguestfs.org/guestfs/_obj//_cgo_.o -gdwarf-2 -c -xc -ferror-limit=0 -Wno-unknown-warning-option -Wno-unneeded-internal-declaration -Wno-unused-function -Qunused-arguments -fno-builtin -I /var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build437084003/libguestfs.org/guestfs/_obj/ -I/private/tmp/libguestfs20151104-69190-1v68vy1/libguestfs-1.30.3/src -DGUESTFS_PRIVATE=1 -m64 -
superenv removed:  -gdwarf-2 -I /var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build437084003/libguestfs.org/guestfs/_obj/ -m64
superenv added:    -pipe -Os -march=native -I/var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build437084003/libguestfs.org/guestfs/_obj/ -isystem/usr/local/include -isystem/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/gettext/include -I/usr/local/opt/readline/include -I/usr/local/opt/libxml2/include
superenv executed: clang -pipe -w -Os -march=native -w -Wno-error -o/var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build437084003/libguestfs.org/guestfs/_obj//_cgo_.o -c -xc -ferror-limit=0 -Wno-unknown-warning-option -Wno-unneeded-internal-declaration -Wno-unused-function -Qunused-arguments -fno-builtin -I/var/folders/1c/382_sr9d6m7d5kqhpr_gt0sh0000gn/T/go-build437084003/libguestfs.org/guestfs/_obj/ -I/private/tmp/libguestfs20151104-69190-1v68vy1/libguestfs-1.30.3/src -DGUESTFS_PRIVATE=1 - -isystem/usr/local/include -isystem/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers -I/usr/local/opt/gettext/include -I/usr/local/opt/readline/include -I/usr/local/opt/libxml2/include
```

**Comment by AntonioMeireles** [comment]
similar issue hit and workarounded in #47405. quoting from there (per @mikemcquaid [request] (https://github.com/Homebrew/homebrew/pull/47405#discussion_r48477728)) here as this looks the proper place to discuss this. 

>FWIW and at sight there's some (`homebrew` env) variable that's passed to `cgo` (when invoked by `godep`, at least) by `make` that's not passed to it when `godep` is called directly by system, and that's what seems to be breaking builds (as the dependencies that are built via `cgo` get stripped prematurely).

>the workaround people seems to be using is either just not using `make` or (as i did in latest iteration) assure that those (go)deps get properly installed before any attempt to consume them.

FWIW also workarounded (by running away from `godep`) in #43073 and in [nomad](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/nomad.rb) formula.

**Comment by mikemcquaid** [comment]
Thanks @AntonioMeireles!

**Comment by mikemcquaid** [comment]
Looks like the `-gdwarf-2` is the problematic bit.

**Comment by DomT4** [comment]
Please let us know if we can assist further here, Thanks! I _think_ it should have been resolved with #47485 being merged.

**Comment by shulima** [comment]
Thanks, will test as soon as I can drag myself away from XCOM2.
